File: Day 3/3-12-2024_Part1.py
import re
def calculate_mul_sum(file_path):
    # Regular expression to match valid mul instructions
    mul_pattern = r"mul\((\d+),(\d+)\)"    
    total_sum = 0

    try:    
        with open(file_path, 'r') as f:
            data = f.read()
            
        # Find all valid mul instructions
        matches = re.findall(mul_pattern, data)
            
        # Calculate the sum of all multiplications
        for match in matches:
            num1, num2 = map(int, match)
            total_sum += num1 * num2
            
    except FileNotFoundError:
        print(f"Error: File not found at {file_path}")
    except Exception as e:
        print(f"An error occurred: {e}")
    
    return total_sum

file_path = "/home/dell/CODING/Advent of Code 2024/Day 3/3-12-2024.txt"
result = calculate_mul_sum(file_path)
print(f"The sum of all valid mul instructions is: {result}")

File: Day 3/3-12-2024_Part1.cpp
#include <iostream>
#include <fstream>
#include <regex>
#include <string>

using namespace std;

int calculateMulSum(const string& filePath) {
    // Regular expression to match valid mul instructions
    regex mulPattern(R"(mul\((\d+),(\d+)\))");
    smatch matches;

    int totalSum = 0;
    string line;

    try {
        ifstream inputFile(filePath);
        if (!inputFile.is_open()) {
            throw ios_base::failure("File not found");
        }

        // Read the file content
        string data((istreambuf_iterator<char>(inputFile)),
                    istreambuf_iterator<char>());

        // Use regex iterator to find all matches
        auto begin = sregex_iterator(data.begin(), data.end(), mulPattern);
        auto end = sregex_iterator();

        for (auto it = begin; it != end; ++it) {
            matches = *it;
            int num1 = stoi(matches[1].str());
            int num2 = stoi(matches[2].str());
            totalSum += num1 * num2;
        }
    } 
    catch (const exception& e) {
        cerr << "An error occurred: " << e.what() << endl;
    }

    return totalSum;
}

int main() {
    string filePath = "/home/dell/CODING/Advent of Code 2024/Day 3/3-12-2024.txt";
    int result = calculateMulSum(filePath);
    cout << "The sum of all valid mul instructions is: " << result << endl;
    return 0;
}